# Введение

Для выполнения лабораторных работ используется Jupyter Notebook. В Jupyter notebook вы можете разрабатывать, документировать и выполнять приложения на языке Python, он состоит из двух компонентов: веб-приложение, запускаемое в браузере, и ноутбуки – файлы, в которых можно работать с исходным кодом программы, запускать его, вводить и выводить данные и т.п.

Веб приложение позволяет:

*    редактировать Python код в браузере, с подсветкой синтаксиса, автоотступами и автодополнением;
*    запускать код в браузере;
*    отображать результаты вычислений с медиа представлением (схемы, графики);
*    работать с языком разметки Markdown и LaTeX.

Ноутбуки – это файлы, в которых сохраняются исходный код, входные и выходные данные

Ноутбук состоит из ячеек (строк). Каждая ячейка может содежать код или текст. Добавляя ячеку вы выбираете какого типа ячейка вам нужна: +Code или +Text. Под кодом мы подразумеваем код, написанный на языке Python или системные команды (для системных команд перед ними ставится !)
В текстовых строках можно ввести любой текст с определенным форматированием, html и математические формулы.



Для начала решим простую арифметическую задачу: добавьте строчку “Code”, введите в ячейке “2 + 3” без кавычек и нажмите Ctrl+Enter или Shift+Enter, в первом случае введенный вами код будет выполнен интерпретатором Python, во втором – будет выполнен код и создана новая ячейка, которая расположится уровнем ниже.

Если ваша программа зависла, то можно прервать ее выполнение выбрав на панели меню пункт Kernel -> Interrupt.

In [1]:
11+3

14

In [2]:
import math

In [3]:
math.e

2.718281828459045

In [4]:
12+3

15

In [5]:
2+3

5

In [6]:
for i in range(10):
    print(i)

0
1
2
3
4
5
6
7
8
9


In [7]:
i = 0
while i < 10:
    print(i)
    i += 1

0
1
2
3
4
5
6
7
8
9


In [8]:
10+1

11

Следующим шагом переназовем ноутбук.
У каждого ноутбука есть имя, оно отображается в верхней части экрана. Для изменения имени нажмите на его текущее имя и поменяйте название ноутбука на Lab1_Фамилия_Группа.ipynb

Дополнительную справочную информацию по работе в ноубуках можно посмотреть в интернете.

#Основы Pandas

Pandas - это библиотека Python, предоставляющая широкие возможности для анализа данных. С ее помощью очень удобно загружать, обрабатывать и анализировать табличные данные с помощью SQL-подобных запросов. В связке с библиотеками Matplotlib и Seaborn появляется возможность удобного визуального анализа табличных данных.


Для использования библиотек их необходимо импортировать, что можно сделать с помощью команды import. После ключевого слова import указывается название библиотеки, а после as - ее краткое название (алиас), которое мы будем использовать в коде для краткого обозначения этой библиотеки.

In [10]:
import pandas as pd

Основными структурами данных в Pandas являются классы Series и DataFrame. Первый из них представляет собой одномерный индексированный массив данных некоторого фиксированного типа (его можно интеерпретировать как индексированный столбец таблицы). Второй - это двухмерная структура данных, представляющая собой таблицу, каждый столбец которой содержит данные одного типа. Можно представлять её как словарь объектов типа Series. Структура DataFrame отлично подходит для представления реальных данных: строки соответствуют признаковым описаниям отдельных объектов, а столбцы соответствуют признакам.

## Чтение из файла и первичный анализ
Чаще всего, данные, которые мы будем анализировать, готовить  и использовать для обучения моделей, не формируются самостоятельно, а берутся из готового дататсета. Рассмотрим чтение датасета (DataFrame) из файла с расширением csv. В качестве примера возьмем titanic_train.csv, который содержить данные о пассажирах Титаника. Полный датасет titanic был разделен на 2 части - тренировочную и тестовую. Именно с тренировочной частью мы и будем работать. Датасет titanic предоставлен в рабочей папке Лаб_1, а также вы можете его скачать (titanic_train.csv) по ссылке https://www.kaggle.com/tedllh/titanic-train#titanic_train.csv. Если вы работаете на платформе Goggle Colab, необходимо сделать upload файла в вашу рабочую директорию ( в меню ноутбука слева  найти и перейти на Files, нажать Upload и выбрать файл для загрузки)

Внимание! Проверьте, что файл находиться в папке content, если его там нет, то нажмите на нее правой кнопкой и upload. 


#Чтение данных

Для чтения csv файла используем функцию pd.read_csv

In [12]:
import pandas as pd
df = pd.read_csv('titanic_train.csv')

Посмотрим тип переменной df

In [13]:
type(df)

pandas.core.frame.DataFrame

### Первичный анализ

После того, как мы считали данные, посмотрим первые несколько записей датафрейма (если в скобках после функции head не указать нужное количество записей, то по умолчанию будут выведены первые 5 записей датафрейма). Если вместо таблицы вы видите  "кашу" из данных, разделенных ";", значит, ваш csv файл использует в качестве разделителя ";", а не ",". И для корректного чтения данных, нужно задать разделитель в функции чтения, например df = pd.read_csv('proba.csv', sep=";")

In [14]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

Давайте посмотрим, что это за данные. Достаточно часто датасеты имеют краткое описание содержимого. Например, что означает тот или иной столбец, какие значения используются для заполнения и т.д. Или значения признаков интуитивно понятны. В нашем случае, мы имеем следующие данные:
каждая строка - это сведения об отдельном пассажире. Строка содержит:
*   PassengerId - идентификатор пассажира (проще говоря, порядковый номер строки)
*   Survived - признак, выжил ли пассажир после крушения (0 - погиб, 1 - выжил)
*   Pclass - в каком классе ехал пассажир (1, 2 или 3)
*   Name - имя пассажира
*   Sex - пол (male - мужской, female - женский)
*   Age - возраст
*   SibSp - количество родственников (по горизонтали - супруг, братья, сестры) на борту
*   Parch - количество родственников (по вертикали - родители, дети) на борту 
*   Ticket - номер билета
*   Fare - цена билета
*   Cabin - номер каюты
*   Embarked - порт посадки (C — Cherbourg, Q — Queenstown, S — Southampton)





Команда shape() показывате размерность датафрейма

Мы видим, что в нашем датафрейме 891 строка (то есть, сведения о 891 пассажире) и 12 столбцов признаков

In [15]:
df.shape

(891, 12)

Команда columns.values() показывате название всех признаков

In [16]:
df.columns.values

array(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'], dtype=object)

Также полезно посмотреть метод info, который показывает краткую статистику по набору данных

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 66.2+ KB


Можно увидеть, что не все признаки у нас заполнены в каждой строке,для некоторых признаков есть пропуски. Например, у нас 891 строка, а Age (возраст) заполнен только в 714 строках. Совсем мало значений в столбце Cabin. 

Есть несколько вариантов работы с такими строками: 
* их можно оставить без изменений, если признак, в котором пропущены данные, малозначим;
* можно удалить все строки, в которых есть пустые значения;
* можно заполнить их какими-либо значениями (0, медианой или любыми другими константами).

Метод info так же показывает какого типа данные в том или ином столбце: цифровые (int или float), текстовые и т.д. В последней строке указан объем памяти, которую занимает датафрейм. При больших объемах данных это позволяет оценить хватит ли вам ресурсов для загрузки в память полного датасета

Еще один полезный метод для начального анализа данных - метод describe(), который выдает краткую статистику по числовым признакам: среднее, стандартное отклонение, минимум, максимум, медианы, квартили и т.д. Самостоятельно посмотрите, что означает каждое понятие. 

In [18]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

##Доступ к элементам данных

###Доступ к элементам DataFrame

Разберемся как посмотреть какие-то конкретные значения в датафрейме. Наш датафрейм представляет собой таблицу, состоящую из строк и столбцов. Обратиться к конкретному столбцу можно используя его имя. Например, выведем первые 10 строк из столбца с возрастом пассажиров

In [19]:
df['Age'].head(10)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5     NaN
6    54.0
7     2.0
8    27.0
9    14.0
Name: Age, dtype: float64

Второй способ обращения - через "."

In [20]:
df.Age.head(10)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5     NaN
6    54.0
7     2.0
8    27.0
9    14.0
Name: Age, dtype: float64

Столбец таблицы - это объек Pandas Series, убедимся в этом определив тип полученного объекта. Этот объект проиндексирован (слева печатаются индексы). В нижней строке есть информацию о его имени и типе значений.
Кроме того, здесь хорошо видны пропуски (незаполненные значения) в некоторых строках (они обозначены NaN - Not-A-Number)

In [21]:
type(df['Age'])

pandas.core.series.Series

Кроме обращения к столбцу по имени, можно использовать индексацию с помощью методов loc и iloc

Метод loc выдает нам подматрицу, используя индексы строк и столбцов. Первый аргумент отвечает за количество записей (как с начала, так и при желании с любого места), а второй - за отраженные в подматрице признаки.


In [22]:
df.loc[0:5, ['Pclass', 'Fare']]

Pclass     Fare
0       3   7.2500
1       1  71.2833
2       3   7.9250
3       1  53.1000
4       3   8.0500
5       3   8.4583

Метод iloc работает схожим образом, за исключением того, что второй аргумент теперь тоже индексы, то есть отражает количество признаков.


In [23]:
df.iloc[0:5, 1:3]

Survived  Pclass
0         0       3
1         1       1
2         1       3
3         1       1
4         0       3

###Доступ к элементам Series

Для примера создадим два объекта Series

In [24]:
s1 = pd.Series(["one", "two", "three", "four", "five"]) # передаем список, индексы проставляются автоматически с 0
print(s1)

0      one
1      two
2    three
3     four
4     five
dtype: object


In [25]:
s2 = pd.Series([1, 2, 3, 4, 5], ['a', 'b', 'c', 'd', 'e'])  # передаем список, индексы заданы символами от 'a' до 'e'
print(s2)

a    1
b    2
c    3
d    4
e    5
dtype: int64


К элементам Series можно обращаться по численному индексу через "[ ]"

In [26]:
s2[2] # второй элемент серии s2 (индексация с 0)

3

К элементам Series можно обращаться по метке:

In [27]:
s2['c']

3

К элементам Series можно применять так называемые slice (срезы) через ':'.
Срезами является часть серии до или после определенного элемента

In [28]:
s2[:2] # все элементы до второго включительно

a    1
b    2
dtype: int64

In [29]:
s2[2:] # второй и все элементы после второго до конца

c    3
d    4
e    5
dtype: int64

In [30]:
s2[2:4] # все элементы со второго по четвертый не включительно. Последний никогда не включается.

c    3
d    4
dtype: int64

In [31]:
s2[:] # все элементы

a    1
b    2
c    3
d    4
e    5
dtype: int64

В поле для индекса можно поместить условное выражение.
Данное выражение можно перевести как вопрос, является ли значение элемента (признака) меньшим или равным двум.

In [32]:
s2[s2 <= 2]

a    1
b    2
dtype: int64

Или даже несколько условных выражений, отделенных "( )". 

Если соединить два условия знаком "&", то результат будет удовлетворять двум условиям, а если знаком "|", то хотя бы одному из них.

In [33]:
s2[(s2 >= 2) & (s2 <=4)]

b    2
c    3
d    4
dtype: int64

In [34]:
s2[(s2 <= 2) | (s2 == 5)]

a    1
b    2
e    5
dtype: int64

Важно, что в условных выражениях равенство пишется как "==" (два знака "равно").

Со структурами Series можно работать как с векторами: складывать, умножать вектор на число и т.п.

In [35]:
s3 = s2 + 10

In [36]:
print(s3)

a    11
b    12
c    13
d    14
e    15
dtype: int64


## Примеры анализа информации

Для примера отберем пассажиров мужского пола, возраст которых меньше сорока лет.

In [37]:
df[(df['Sex'] == 'male') & (df.Age < 40)].head(5)

PassengerId  Survived  Pclass                            Name   Sex   Age  \
0             1         0       3         Braund, Mr. Owen Harris  male  22.0   
4             5         0       3        Allen, Mr. William Henry  male  35.0   
7             8         0       3  Palsson, Master. Gosta Leonard  male   2.0   
12           13         0       3  Saundercock, Mr. William Henry  male  20.0   
13           14         0       3     Andersson, Mr. Anders Johan  male  39.0   

    SibSp  Parch     Ticket    Fare Cabin Embarked  
0       1      0  A/5 21171   7.250   NaN        S  
4       0      0     373450   8.050   NaN        S  
7       3      1     349909  21.075   NaN        S  
12      0      0  A/5. 2151   8.050   NaN        S  
13      1      5     347082  31.275   NaN        S

Можно отсортировать этих людей по возрастанию платы за билет. Ascending в данной функции отвечает за сортировку по убыванию (False) или по возрастанию (True).

In [38]:
df[(df['Sex'] == 'male') & (df.Age < 40)].sort_values(by='Fare',ascending=True).head(10)

PassengerId  Survived  Pclass                             Name   Sex  \
271          272         1       3     Tornquist, Mr. William Henry  male   
822          823         0       1  Reuchlin, Jonkheer. John George  male   
806          807         0       1           Andrews, Mr. Thomas Jr  male   
179          180         0       3              Leonard, Mr. Lionel  male   
302          303         0       3  Johnson, Mr. William Cahoone Jr  male   
378          379         0       3              Betros, Mr. Tannous  male   
872          873         0       1         Carlsson, Mr. Frans Olof  male   
843          844         0       3       Lemberopolous, Mr. Peter L  male   
371          372         0       3        Wiklund, Mr. Jakob Alfred  male   
202          203         0       3       Johanson, Mr. Jakob Alfred  male   

      Age  SibSp  Parch   Ticket    Fare        Cabin Embarked  
271  25.0      0      0     LINE  0.0000          NaN        S  
822  38.0      0      0    19972  0.0000          NaN        S  
806  39.0      0      0   112050  0.0000          A36        S  
179  36.0      0      0     LINE  0.0000          NaN        S  
302  19.0      0      0     LINE  0.0000          NaN        S  
378  20.0      0      0     2648  4.0125          NaN        C  
872  33.0      0      0      695  5.0000  B51 B53 B55        S  
843  34.5      0      0     2683  6.4375          NaN        C  
371  18.0      1      0  3101267  6.4958          NaN        S  
202  34.0      0      0  3101264  6.4958          NaN        S

Можно подсчитать количество пассажиров (а точнее их уникальных номеров) с помощью функции count()

In [39]:
df[(df['Sex'] == 'male') & (df.Age < 40)].sort_values(by='Fare',ascending=True).PassengerId.count()

344

Также можно быстро вывести распределение переменной Sex с помощью функции value_count()

In [40]:
df.Sex.value_counts()

male      577
female    314
Name: Sex, dtype: int64

## Задания

Задание 1.
Сколько человек спаслись и сколько погибли?

In [41]:
df[(df.Survived == 1)].PassengerId.count()

342

In [42]:
df[(df.Survived == 0)].PassengerId.count()

549

Задание 2. Выведите распределение переменной Survived отдельно для мужчин и отдельно для женщин


In [43]:
df[(df['Sex'] == 'male')].Survived.value_counts()

0    468
1    109
Name: Survived, dtype: int64

In [44]:
df[(df['Sex'] == 'female')].Survived.value_counts()

1    233
0     81
Name: Survived, dtype: int64

Задание 3. Каковы медиана и стандартное отклонение цен за билет?

In [45]:
df.Fare.median()

14.4542

In [46]:
df.Fare.std()

49.6934285971809

In [47]:
df.Fare.median(),\
df.Fare.std()

(14.4542, 49.6934285971809)

Задание 4. Подсчитайте долю выживших среди людей моложе 25 лет и старше 55.

In [52]:
df[(df.Survived == 1) & (df.Age > 55)].PassengerId.count()/df[(df.Age > 55)].PassengerId.count() * 100

30.0

In [53]:
df[(df.Survived == 1) & (df.Age < 25)].PassengerId.count()/df[(df.Age < 25)].PassengerId.count() * 100

42.44604316546763

Задание 5. Подсчитайте долю выживших среди мужчин и женщин.

In [54]:
df[(df.Survived == 1) & (df.Sex == 'male')].PassengerId.count()/df[(df.Sex == 'male')].PassengerId.count() * 100

18.890814558058924

In [55]:
df[(df.Survived == 1) & (df.Sex == 'female')].PassengerId.count()/df[(df.Sex == 'female')].PassengerId.count() * 100

74.20382165605095

Задание 6. В каком городе больше всего село пассажиров?

In [56]:
df.Embarked.value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

Задание 7. Подсчитайте средний возраст выживших

In [57]:
df[(df.Survived == 1)].Age.mean()

28.343689655172415

Задание 8. Подсчитайте средний возраст пассажиров каждого из трех классов


In [58]:
df[(df.Pclass == 1)].Age.mean()

38.233440860215055

In [59]:
df[(df.Pclass == 2)].Age.mean()

29.87763005780347

In [60]:
df[(df.Pclass == 3)].Age.mean()

25.14061971830986

Задание 9. Подсчитайте средний возраст выживших мужчин, купивших билет во 2 класс

In [61]:
df[(df.Pclass == 2)&(df.Sex == 'male')].Age.mean()

30.74070707070707

Задание 10*. Оцените, как влияет на выживаемость пассажиров наличие их родственников на борту.

In [67]:
df[(df.Survived == 1)].SibSp.mean()

0.47368421052631576

In [68]:
df[df.Survived == 1].SibSp.max()

4

In [69]:
df[df.Survived == 1].SibSp.min()

0